robottelo/cli/factory.py

```diff
@@ -3635,3 +3715,338 @@ def add_role_permissions(role_id, resource_permissions):
+        virtual machine to connect to hypervisor
+    :param str subscription_name: the subscription name to assign to virt-who
+        hypervisor guests
+    :param bool exec_one_shot: whether to run the virt-who one-shot command
+        after startup
+    :param bool upload_manifest: whether to upload the organization manifest
+    """
+    if org_id is None:
+        org = make_org()
+    else:
+        org = Org.info({'id': org_id})
+
+    if lce_id is None:
+        lce = make_lifecycle_environment({'organization-id': org['id']})
+    else:
+        lce = LifecycleEnvironment.info({
+            'id': lce_id,
+            'organization-id': org['id']
+        })
+    if upload_manifest:
+        # Upload the org manifest
+        try:
+            manifests.upload_manifest_locked(org_id, manifests.clone(),
+                                             interface=manifests.INTERFACE_CLI)
+        except CLIReturnCodeError as err:
+            raise CLIFactoryError(
+                u'Failed to upload manifest\n{0}'.format(err.msg))
+    # setup repositories
+    cdn_repos = []
+    # Red Hat Enterprise Linux 7 Server
+    rh_product_arch = REPOS['rhel7']['arch']
+    rh_product_releasever = REPOS['rhel7']['releasever']
+    cdn_repos.append({
+        'product': PRDS['rhel'],
+        'repository-set': REPOSET['rhel7'],
+        'repository': REPOS['rhel7']['name'],
+        'repository-id': REPOS['rhel7']['id'],
+        'releasever': rh_product_releasever,
+        'arch': rh_product_arch
+    })
+    # Red Hat Satellite Tools 6.2 for RHEL 7
+    if settings.cdn or not settings.sattools_repo['rhel7']:
+        cdn_repos.append({
+            'product': PRDS['rhel'],
+            'repository-set': REPOSET['rhst7'],
+            'repository': REPOS['rhst7']['name'],
+            'repository-id': REPOS['rhst7']['id'],
+        })
+    # # Enable the cdn RH products
+    for repo in cdn_repos:
+        try:
+            RepositorySet.enable({
+                u'basearch': rh_product_arch,
+                u'name': repo['repository-set'],
+                u'organization-id': org['id'],
+                u'product': repo['product'],
+                u'releasever': repo.get('releasever'),
+            })
+        except CLIReturnCodeError as err:
+            raise CLIFactoryError(
+                u'Failed to enable repository set\n{0}'.format(err.msg))
+    # Retrieve the repositories info
+    repos_info = []
+    for repo in cdn_repos:
+        try:
+            repo_info = Repository.info({
+                u'name': repo['repository'],
+                u'organization-id': org['id'],
+                u'product': repo['product'],
+            })
+            repos_info.append(repo_info)
+        except CLIReturnCodeError as err:
+            raise CLIFactoryError(
+                u'Failed to fetch repository info\n{0}'.format(err.msg))
+    custom_product = None
+    # If we aren't working with cdn, create custom repo
```

Comment:
Hmm. On that step I started to dislike current approach
1) Having such long methods is a bad practice as for me. I stop to understand general idea to the end.
Suggestion: If it is possible - divide to sub-methods by logical parts
2) It doesn't seem correct to keep both cdn and custom repos handlers in one method as in one case some parts of the code will be skipped and in second almost 50% of code.
Suggestion: if we go with 1) and divide code on some parts - we can call for necessary sub-methods depending on conditions
3) You started to use a lot of exception handlers which is unusual for our code. Basically, I don't mind as it makes code better, but not in terms of size and readability. 
Suggestion: Maybe we can create small handler on a top level that will accept statement and message or something like this?